processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -133,13 +168,16 @@ protected AddToFactsResult addToFacts(
 
       // Last ditch sanity checks
       if (numEntries.get() >= maxRowCount
-          && facts.getPriorIndex(key) == TimeAndDims.EMPTY_ROW_INDEX
-          && !skipMaxRowsInMemoryCheck) {
-        throw new IndexSizeExceededException("Maximum number of rows [%d] reached", maxRowCount);
+          || maxBytesInMemory != -1 && sizeInBytes.get() >= maxBytesInMemory
```

Comment:
Please include some clarifying parentheses here. I can never remember how mixed and/or code like `a || b && c && d` is meant to be evaluated and I don't think I'm alone in this.

Reply:
done